src/core/server/elasticsearch/client/cluster_client.ts

```diff
@@ -0,0 +1,119 @@
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Client } from '@elastic/elasticsearch';
+import { Logger } from '../../logging';
+import { GetAuthHeaders, isRealRequest, Headers } from '../../http';
+import { ensureRawRequest, filterHeaders } from '../../http/router';
+import { ScopeableRequest } from '../types';
+import { ElasticSearchClient } from './types';
+import { configureClient } from './configure_client';
+import { ElasticsearchClientConfig } from './client_config';
+import { ScopedClusterClient, IScopedClusterClient } from './scoped_cluster_client';
+
+const noop = () => undefined;
+
+/**
+ * Represents an Elasticsearch cluster API client created by the platform.
+ * It allows to call API on behalf of the internal Kibana user and
+ * the actual user that is derived from the request headers (via `asScoped(...)`).
+ *
+ * @public
+ **/
+export interface IClusterClient {
+  /**
+   * Returns a {@link ElasticSearchClient | client} to be used to query the ES cluster on behalf of the Kibana internal user
+   */
+  asInternalUser: () => ElasticSearchClient;
+  /**
+   * Creates a {@link IScopedClusterClient | scoped cluster client} bound to given {@link ScopeableRequest | request}
+   */
+  asScoped: (request: ScopeableRequest) => IScopedClusterClient;
+}
+
+/**
+ * See {@link IClusterClient}
+ *
+ * @public
+ */
+export interface ICustomClusterClient extends IClusterClient {
+  /**
+   * Closes the cluster client. After that client cannot be used and one should
+   * create a new client instance to be able to interact with Elasticsearch API.
+   */
+  close: () => void;
+}
+
+/** @internal **/
+export class ClusterClient implements IClusterClient, ICustomClusterClient {
+  private readonly internalClient: Client;
+  private readonly rootScopedClient: Client;
+
+  private isClosed = false;
+
+  constructor(
+    private readonly config: ElasticsearchClientConfig,
+    logger: Logger,
+    private readonly getAuthHeaders: GetAuthHeaders = noop
+  ) {
+    this.internalClient = configureClient(config, { logger });
+    this.rootScopedClient = configureClient(config, { logger, scoped: true });
+  }
+
+  asInternalUser() {
+    return this.internalClient;
+  }
+
+  asScoped(request: ScopeableRequest) {
+    const scopedHeaders = this.getScopedHeaders(request);
+    const scopedClient = this.rootScopedClient.child({
+      headers: scopedHeaders,
+    });
+    return new ScopedClusterClient(this.internalClient, scopedClient);
+  }
+
+  public close() {
+    if (this.isClosed) {
+      return;
+    }
+
+    this.isClosed = true;
+    this.internalClient.close();
+    this.rootScopedClient.close();
+  }
```

Comment:
The `client.close` method returns a promise, if you are not interested in waiting for its resolution I would at least add a `catch` handler *(even if we don't expect errors here)* to avoid unhandled rejections :)

Reply:
missed that it was async. Will address.

Reply:
https://github.com/elastic/kibana/pull/69905/commits/a764b10db4343b0ac799eb4260aa7dd4baccf390